README
If you like my work, buy me a drink https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=ZWE8DKMCP6ENJ


CHANGELOG
1.0.0 (20140702)
  -First stable version
  -Added auto numbering
  -Fixed some bugs
0.0.0 (20130610)
  -Initial release


lib/thepyutilities/__init__.py
#!/usr/bin/env python
###############################################################################################
#  Author:
_author = '<a href="mailto:debuti@gmail.com">Borja Garcia</a>'
# Program:
_name = 'thepyutilities'
# Descrip:
_description = '''Library'''
# Version:
_version = '1.1'
#    Date: 
_date = '2010-10-29'
# License: This script doesn't require any license since it's not intended to be redistributed.
#          In such case, unless stated otherwise, the purpose of the author is to follow GPLv3.
# History: 1.1 (2010-10-29)
#            -Added several methods and several bug fixed
# History: 1.0 (2009-03-09)
#            -Initial release
###############################################################################################

__all__ = ["shellutils", "stringutils", "filesystemutils", "generalutils"]


lib/thepyutilities/timeutils.py
#!/usr/bin/env python
###############################################################################################
#  Author:
__author__ = '<a href="mailto:debuti@gmail.com">Borja Garcia</a>'
# Program:
__program__ = 'timeutils'
# Package:
#__package__ = 'tk.ganian.thepyutilities'
# Descrip:
__description__ = '''Library for time utilities.'''
# Version:
__version__ = '0.0.1'
#    Date: 
__date__ = '2012-04-24'
# License: This script doesn't require any license since it's not intended to be redistributed.
#          In such case, unless stated otherwise, the purpose of the author is to follow GPLv3.
# History:
#           0.0.1 (2012-04-24)
#            -Initial release
###############################################################################################

import subprocess, time
import os
import socket
import logging
import sys
import re
import shutil
import stringutils
import hashlib






lib/thepyutilities/stringutils.py
#!/usr/bin/env python
###############################################################################################
#  Author:
_author = '<a href="mailto:debuti@gmail.com">Borja Garcia</a>'
# Program:
_name = 'stringutils'
# Descrip:
_description = '''Library for string related utilities.'''
# Version:
_version = '0.0.2'
#    Date: 
_date = '2010-10-29'
# License: This script doesn't require any license since it's not intended to be redistributed.
#          In such case, unless stated otherwise, the purpose of the author is to follow GPLv3.
# History:
#          0.0.2 (2010-10-29)
#            -Fixed some errors
#          0.0.1 (2009-11-27)
#            -Initial release
###############################################################################################

import subprocess
import os
import socket
import logging
import string
import sys
import re
import shutil

def extendedFind(stringToAnalize, substringToSearch, startPosition, endPosition, numberOfCoincidence = 1):
    '''Finds a substring in a string, like find, but adds support to search\
       for the second or consecutive coincidences. If not found return the last encountered
    '''
    coincidencesCounted = 0
    lastCoincidencePosition = startPosition

    while coincidencesCounted < numberOfCoincidence:
        index = stringToAnalize.find(substringToSearch, lastCoincidencePosition, endPosition)
        #print "Busco desde " + str(lastCoincidencePosition) + " hasta " + str(endPosition) + " y encuentro el salto de linea en la posicion " + str(index)
        if index <= startPosition or index >= endPosition: 
            break
        else:
            lastCoincidencePosition = index + 1
            coincidencesCounted = coincidencesCounted + 1

    if coincidencesCounted != numberOfCoincidence:
        return lastCoincidencePosition
    else:
        return index

def extendedRFind(stringToAnalize, substringToSearch, startPosition, endPosition, numberOfCoincidence = 1):
    '''Finds a substring in a string, like find, but adds support to search\
       for the second or consecutive coincidences. If not found return the last encountered
    '''
    coincidencesCounted = 0
    lastCoincidencePosition = endPosition
    
    while coincidencesCounted < numberOfCoincidence:
        index = stringToAnalize.rfind(substringToSearch, startPosition, lastCoincidencePosition)
        #print "Busco desde " + str(startPosition) + " hasta " + str(lastCoincidencePosition) + " y encuentro el salto de linea en la posicion " + str(index)
        if index <= startPosition or index >= endPosition: 
            break
        else:
            lastCoincidencePosition = index - 1
            coincidencesCounted = coincidencesCounted + 1
     
    if coincidencesCounted != numberOfCoincidence:
        return lastCoincidencePosition
    else:
        return index


def list2str(list):
    '''This function turn a list into a str
    '''
    if type(list).__name__ != 'list':
        return None
    else:
        result = ""
        for element in list:
            result += element
            result += "\n"
        return result

lib/thepyutilities/generalutils.py
#!/usr/bin/env python
###############################################################################################
#  Author:
_author = '<a href="mailto:debuti@gmail.com">Borja Garcia</a>'
# Program:
_name = 'generalutils'
# Descrip:
_description = '''Library for uncataloged utilities.'''
# Version:
_version = '0.0.2'
#    Date: 
_date = '2010-11-05'
# License: This script doesn't require any license since it's not intended to be redistributed.
#          In such case, unless stated otherwise, the purpose of the author is to follow GPLv3.
# History:
#          0.0.1 (2010-11-05)
#            -Initial release
###############################################################################################

import os
import time
import shellutils
import socket
import platform

import smtplib
from email.MIMEMultipart import MIMEMultipart
from email.MIMEBase import MIMEBase
from email.MIMEText import MIMEText
from email import Encoders


def osName():
    '''This function returns the os name
    '''
    if (platform.system() == 'Linux'):
        (distname,version,id) = platform.linux_distribution()
        return distname
    return platform.system()


def osVersion():
    '''This function returns the os version
    '''
    if (platform.system() == 'Linux'):
        #platform.linux_distribution() output will be something like this ('Ubuntu', '12.04', 'precise')
        return platform.linux_distribution()[1]
    if (platform.system() == 'Windows'):
        return platform.system()
    return None

    return platform.release()


def systemName():
    '''This function returns the system name
    '''
    return socket.gethostname()
    

def now():
    '''This function returns the current time
    '''
    return time.strftime("%Y-%m-%d:%H:%M:%S")


def isInCron(line):
    '''This method checks if crontab contains one line 
    '''
    code, output, error = shellutils.run(["crontab", "-l"])
     
    if len(error) != 0 or shellutils.grep(line, output) == None:
        return False
        
    else:
        return True


def setCron(line):
    '''This method add a line to crontab
    '''
    TEMP_FILE="/tmp/crontab.edit.tmp"
    
    code, output, error = shellutils.run(["crontab", "-l"])
    if len(error) == 0:
        file_out = open(TEMP_FILE, 'w')
        file_out.write(output + "\n" + line + "\n")
        file_out.close()
        code, output, error = shellutils.run(["crontab", TEMP_FILE])
        if len(error) == 0:
            return 0
        else:
            return None
    else:
        return None


def mail(gmailUser, gmailPwd, to, subject, text, attach=None):
    '''This procedure sends an email to a list of recipients
    '''
    msg = MIMEMultipart()

    msg['From'] = gmailUser
    msg['To'] = to
    msg['Subject'] = subject

    msg.attach(MIMEText(text))

    if attach != None:
        part = MIMEBase('application', 'octet-stream')
        part.set_payload(open(attach, 'rb').read())
        Encoders.encode_base64(part)
        part.add_header('Content-Disposition',
           'attachment; filename="%s"' % os.path.basename(attach))
        msg.attach(part)

    mailServer = smtplib.SMTP("smtp.gmail.com", 587)
    mailServer.ehlo()
    mailServer.starttls()
    mailServer.ehlo()
    mailServer.login(gmailUser, gmailPwd)
    mailServer.sendmail(to, to, msg.as_string())
    # Should be mailServer.quit(), but that crashes...
    mailServer.close()


def notifyPushover(appId, userId, message, title="Python app"):
    '''This function is for sending push notifications
    '''
    import httplib, urllib
    conn = httplib.HTTPSConnection("api.pushover.net:443")
    conn.request("POST", "/1/messages.json",
      urllib.urlencode({
        "token": appId,
        "user": userId,
        "message": message,
		"title": title,
      }), { "Content-type": "application/x-www-form-urlencoded" })
    conn.getresponse()


lib/thepyutilities/filesystemutils.py
#!/usr/bin/env python
###############################################################################################
#  Author:
_author = '<a href="mailto:debuti@gmail.com">Borja Garcia</a>'
# Program:
_name = 'filesystemutils'
# Descrip:
_description = '''Library for file system utilities.'''
# Version:
_version = '0.0.1'
#    Date: 
_date = '2012-04-28'
# License: This script doesn't require any license since it's not intended to be redistributed.
#          In such case, unless stated otherwise, the purpose of the author is to follow GPLv3.
# History:
#          0.0.1 (2012-04-28)
#            -Initial release
###############################################################################################


def removeLine(filename, lineno):
    '''This function returns one line of a file and removes it
    '''
    fro = open(filename, "rb")

    current_line = 0
    while current_line < lineno:
        fro.readline()
        current_line += 1

    seekpoint = fro.tell()
    frw = open(filename, "r+b")
    frw.seek(seekpoint, 0)

    # read the line we want to discard
    readedLine = fro.readline()

    # now move the rest of the lines in the file 
    # one line back 
    chars = fro.readline()
    while chars:
        frw.writelines(chars)
        chars = fro.readline()

    fro.close()
    frw.truncate()
    frw.close()
    
    return readedLine


src/autotag.py
#!/usr/bin/env python
###############################################################################################
#  Author: 
__author__ = '<a href="mailto:debuti@gmail.com">Borja Garcia</a>'
# Program: 
__program__ = 'autotag'
# Package:
__package__ = ''
# Descrip: 
__description__ = '''CLI application for tagging and guessing tags on music files. Also implements 
a pattern renamer.'''
# Version: 
__version__ = '1.0.0'
#    Date:
__date__ = '20140702'
# License: This script doesn't require any license since it's not intended to be redistributed.
#          In such case, unless stated otherwise, the purpose of the author is to follow GPLv3.
# History: 
#          1.0.0 (20140702)
#            -First stable version
#            -Added auto numbering
#            -Fixed some bugs
#          0.0.0 (20130610)
#            -Initial release
###############################################################################################

# Imports
import logging
import sys
import doctest
import datetime, time
import os
import optparse
import inspect
import ConfigParser
import glob
import traceback
import json
import urllib

# Parameters, Globals n' Constants
KIBI = 1024
MEBI = 1024 * KIBI
LOG_MODE = "Screen"
LOG_LEVEL = logging.DEBUG
LOG_MAX_BYTES = 1 * MEBI

APIKEY = "TKTD9246TZYXHOE2P"

realScriptPath = os.path.realpath(__file__)
realScriptDirectory = os.path.dirname(realScriptPath)
callingDirectory = os.getcwd()
if os.path.isabs(__file__ ):
    linkScriptPath = __file__
else:
    linkScriptPath = os.path.join(callingDirectory, __file__)
linkScriptDirectory = os.path.dirname(linkScriptPath)

propertiesName = __program__ + ".properties"
propertiesPath = os.path.join(realScriptDirectory, '..', propertiesName) 

logFileName = __program__ + '_' + time.strftime("%Y%m%d%H%M%S") + '.log'
logDirectory = os.path.join(realScriptDirectory, '..', 'logs')
logPath = os.path.join(logDirectory, logFileName)
loggerName = __package__ + "." + __program__

# User-libs imports (This is the correct way to do this)
libPath =  os.path.join(realScriptDirectory, '..', 'lib')
sys.path.insert(0, libPath)
for infile in glob.glob(os.path.join(libPath, '*.*')):
    sys.path.insert(0, infile)
    
import thepyutilities.shellutils as shellutils


# Usage function, logs, utils and check input
def openLog(mode, desiredLevel):
    '''This function is for initialize the logging job
    '''
    def openScreenLog(formatter, desiredLevel):
        logging.basicConfig(level = desiredLevel, format = formatter)
       
    def openScreenAndFileLog(fileName, formatter, desiredLevel):
        logger = logging.getLogger('')
        logger.setLevel(desiredLevel)
        # create file handler which logs even debug messages
        fh = logging.FileHandler(fileName)
        fh.setLevel(desiredLevel)
        fh.setFormatter(formatter)
        # add the handler to logger
        logger.addHandler(fh)

    def openScreenAndRotatingFileLog(fileName, formatter, desiredLevel, maxBytes):
        logger = logging.getLogger('')
        logger.setLevel(desiredLevel)
        # create file handler which logs even debug messages
        fh = logging.handlers.RotatingFileHandler(fileName, maxBytes)
        fh.setLevel(desiredLevel)
        fh.setFormatter(formatter)
        # add the handler to logger
        logger.addHandler(fh)

    format = "%(asctime)-15s - %(levelname)-6s - %(funcName)10.10s - %(message)s"
    formatter = logging.Formatter(format)
    # Clean up root logger
    for handler in logging.getLogger('').handlers:
        logging.getLogger('').removeHandler(handler)
    openScreenLog(format, desiredLevel)
    
    if mode == "File" or mode == "RollingFile":
        if not os.path.isdir(logDirectory):
            shellutils.mkdir(logDirectory)
  
        if mode == "File":
            openScreenAndFileLog(logPath, formatter, desiredLevel)
    
        elif mode == "RollingFile":
            openScreenAndRotatingFileLog(logPath, formatter, desiredLevel, LOG_MAX_BYTES)

def closeLog():
    '''This function is for shutdown the logging job
    '''
    logging.shutdown()

def checkInput():
    '''This function is for treat the user command line parameters.
    '''
    # Custom parser to show epilog info
    class MyParser(optparse.OptionParser):
        def format_epilog(self, formatter):
            return self.epilog

    examples = """
 Examples:
  - Make a custom album
     autotag -N -s -r %n\ -\ %a\ -\ %t -A My\ album -Y 2014 -T Various\ Artists <dir>\*mp3
  - Make BorjueloMix
     autotag -N -s -r %n\ -\ %a\ -\ %t -A BorjueloMix2011 -Y 2011 -T Various\ Artists /media/dropbox/Queues/nettop/Music/youtubeSingles/2011/*mp3
"""

    # Create instance of OptionParser Module, included in Standard Library
    p = MyParser(description=__description__,
                 prog=__program__,
                 version=__version__,
                 usage='''%prog [options] <files>''', 
                 epilog=examples) 

    # Define the options. Do not use -h nor -v, the are reserved to help and version automaticly
    p.add_option('--rename', '-r', action="store", type="string", dest="rename", help='If setted will rename to regexp. The wildcards matches the program options. Ex. "%a - %t" will rename to "artist - title"')
    p.add_option('--auto-track-num','-N', action="store_true", dest="autotracknum", help='Auto track number: Use a consecutive number for each file (Caution, this superseed -n)')
    p.add_option('--dont-erase', '-d', action="store_true", dest="donterase", help='If setted eyed3 will not call \"--remove-all\"')
    p.add_option('--ask', '-s', action="store_true", dest="ask", help='Ask for confirmation before doing anything')
    
    p.add_option('--artist', '-a', action="store", type="string", dest="artist", help='The artist. Overwritten if the analysis returns something useful')
    p.add_option('--title', '-t', action="store", type="string", dest="title", help='The title. Overwritten if the analysis returns something useful')
    p.add_option('--album','-A', action="store", type="string", dest="album", help='The album')
    p.add_option('--track-num','-n', action="store", type="int", dest="tracknum", help='The track number')
    p.add_option('--year','-Y', action="store", type="int", dest="year", help='The year')
    p.add_option('--albumartist','-T', action="store", type="string", dest="albumartist", help='The album artist')
    
    # Parse the commandline
    options, arguments = p.parse_args()

    # Decide what to do
    if len(arguments) == 0:
        p.print_help()
        sys.exit(-1)
    else:
        # Return as [properties, options, files]
        return [{'artist':options.artist, 
                 'title':options.title, 
                 'album':options.album, 
                 'tracknum':options.tracknum, 
                 'year':options.year,
                 'albumartist':options.albumartist,
                }, 
                {'rename':options.rename,
                 'donterase':options.donterase,
                 'ask':options.ask,
                 'autotracknum':options.autotracknum,
                },
                arguments]
    

# Helper functions
def areToolsInstalled():
    '''
    '''
    result = True
    softwareNeeded = [{'sw':"echoprint-codegen", 
                       'message':"""Tool needed not found. Install it by typing: 
 sudo apt-get install ffmpeg libboost-all-dev libtag1-dev zlib1g-dev git
 git clone -b release-4.12 http://github.com/echonest/echoprint-codegen.git
 cd echoprint-codegen/
 cd src/
 make
 cd ..
 cd ..
 mv echoprint-codegen /usr/local/.
 ln -s /usr/local/echoprint-codegen/echoprint-codegen /usr/bin/echoprint-codegen"""},
                      {'sw':"eyeD3", 'message':"""Tool needed not found. Install it by typing: 
 sudo apt-get install eyed3"""},
]
    for software in softwareNeeded:
        if not shellutils.executableExists(software['sw']):
            print software['message']
            result = False
            
    return result

def echoprint(audiofile):
    '''
    '''
    artist = None
    title = None
    
    logging.info("Looking up: " + audiofile)
    
    # Run echoprint
    command = ["echoprint-codegen",
                audiofile,
                "10",
                "30"];
    status, stdout, stderr = shellutils.run(command);
    stdoutDecoded = json.loads(stdout)
    if len(stdoutDecoded) > 0 and 'code' in stdoutDecoded[0]:
        code = stdoutDecoded[0]['code']
        
        # Query server
        url = "http://developer.echonest.com/api/v4/song/identify?api_key=%s&code=%s&version=%s" % (urllib.quote(APIKEY), urllib.quote(code), urllib.quote("4.12"))
        #logging.debug("URL:    " + url)
        webout = urllib.urlopen(url).read()
        #logging.debug("Result: " + webout)
        
        weboutDecoded = json.loads(webout)    
        status = weboutDecoded['response']['status']['code']
        if status == 0 and len(weboutDecoded['response']['songs']) > 0:
            artist = weboutDecoded['response']['songs'][0]['artist_name']
            title =  weboutDecoded['response']['songs'][0]['title']
        else:
            logging.info("Nothing found for: " + audiofile)
    else:
        logging.error("  Error analyzing: " + stdout)   
        
    return [artist, title]

def writeTags(properties, options, audiofile):
    ''' Write tags with eyed3
    '''
    def escapeCharacters(input):
        return str(input)
    
    # Clear tags and write new ones
    command = ["eyeD3"]
    if options['donterase'] == None:
        command.append("--remove-all")
    if properties['artist'] != None:
        command.append("-a")
        command.append(escapeCharacters(properties['artist']))
    if properties['title'] != None:
        command.append("-t")
        command.append(escapeCharacters(properties['title']))
    if properties['album'] != None:
        command.append("-A")
        command.append(escapeCharacters(properties['album']))
    if properties['tracknum'] != None:
        command.append("-n")
        command.append(str(properties['tracknum']))
    if properties['year'] != None:
        command.append("-Y")
        command.append(str(properties['year']))
    if properties['albumartist'] != None:
        command.append("--set-user-text-frame=ALBUMARTISTSORT:" + escapeCharacters(properties['albumartist']))
    command.append(audiofile)
                
    logging.debug("  %s" % " ".join(map(str, command)));
    status, stdout, stderr = shellutils.run(command);
    if status == 0:
        for key in properties.keys():
            if properties[key] != None:
                logging.info("  "+key+": \t" + str(properties[key]))
        logging.info("File modified: " + audiofile)

def rename(properties, options, audiofile):
    '''
    '''
    def escapeCharacters(input):
        return input.replace("'", "")
    
    # Import the regular expression
    newname = options['rename']

    # TODO: Check if something would remain unchanged in the filename
    # Substitute every item
    newname = newname.replace("%a", escapeCharacters(properties['artist']))
    newname = newname.replace("%t", escapeCharacters(properties['title']))
    newname = newname.replace("%A", escapeCharacters(properties['album']))
    newname = newname.replace("%n", str(properties['tracknum']))
    newname = newname.replace("%Y", str(properties['year']))
    newname = newname.replace("%T", str(properties['albumartist']))
    newname = os.path.join(shellutils.dirname(audiofile), newname + shellutils.extension(audiofile))
    shellutils.mv(audiofile, newname)
    logging.info("File renamed to: " + newname)

    
# Main function
def core(properties, options, files):
    '''This is the core, all program logic is performed here
    '''
    for index, audiofile in enumerate(files):
        # Do dict initialization
        values = {}

        [fetchedArtist, fetchedTitle] = echoprint(audiofile)
                
        values['artist'] = fetchedArtist or properties['artist']
        values['title'] = fetchedTitle or properties['title']
        values['album'] = properties['album']
        if options['autotracknum']:
            values['tracknum'] = "%03d" % (index + 1)
        else:
            values['tracknum'] = properties['tracknum']
        values['albumartist'] = properties['albumartist']
        values['year'] = properties['year']
        
        if values['album'] == None and \
           values['title'] == None and \
           values['artist'] == None and \
           values['tracknum'] == None and \
           values['albumartist'] == None and \
           values['year'] == None:
           
            logging.info("File not modified: " + audiofile)
            
        else:
            if options['ask'] != None:
                print "I'm going to set this tags to this file " + audiofile
                for key in values.keys():
                    if values[key] != None:
                        print "  "+key+": \t" + str(values[key])
                if options['rename'] != None:
                    print "And I'm going to rename it too."
                print "Do you want it? (Y/n)"
                ok = raw_input().lower()
                if ok is not None and ok == 'n':
                    print "Ok, i won't do that.."
                    continue
                
            writeTags(values,
                      options,
                      audiofile)
         
            if options['rename'] != None:
                rename(values,
                       options, 
                       audiofile)

def main():
    '''This is the main procedure, is detached to provide compatibility with the updater
    '''
    openLog(LOG_MODE, LOG_LEVEL)
    [properties, options, files] = checkInput()
    logging.debug(str([properties, options, files]))
    core(properties, options, files)
    closeLog()

    
# Entry point
if __name__ == '__main__':
    try:
        if areToolsInstalled():
            main()
    
    except KeyboardInterrupt:
        print "Shutdown requested. Exiting"
    except SystemExit:
        pass
    except:
        logging.error("Unexpected error:" + traceback.format_exc())
        raise
